File: app/src/main/java/com/example/myapplication/data/local/WorkoutFeedbackDao.kt
package com.example.myapplication.data.local

import androidx.room.Dao
import androidx.room.Query
import androidx.room.Upsert
import kotlinx.coroutines.flow.Flow

@Dao
interface WorkoutFeedbackDao {
    @Query(
        """
        SELECT * FROM workout_feedback
        WHERE goalId = :goalId
        ORDER BY completedEpochDay DESC, recordedAtEpochMillis DESC
        """,
    )
    fun observeForGoal(goalId: Long): Flow<List<WorkoutFeedbackEntity>>

    @Query("SELECT * FROM workout_feedback WHERE sessionId = :sessionId LIMIT 1")
    suspend fun feedbackForSessionNow(sessionId: Long): WorkoutFeedbackEntity?

    @Upsert
    suspend fun upsert(feedback: WorkoutFeedbackEntity)
}

File: flutter/lib/data/local/daos/workout_feedback_dao.dart
import 'package:drift/drift.dart';
import '../database.dart';
import '../tables/workout_feedback.dart';
import '../tables/workout_sessions.dart';

part 'workout_feedback_dao.g.dart';

@DriftAccessor(tables: [WorkoutFeedbacks, WorkoutSessions])
class WorkoutFeedbackDao extends DatabaseAccessor<GymDatabase>
    with _$WorkoutFeedbackDaoMixin {
  WorkoutFeedbackDao(super.db);

  Stream<List<WorkoutFeedbackData>> observeForGoal(int goalId) {
    final query = select(workoutFeedbacks)
      ..where((tbl) => tbl.goalId.equals(goalId))
      ..orderBy([
        (tbl) => OrderingTerm.desc(tbl.completedEpochDay),
        (tbl) => OrderingTerm.desc(tbl.recordedAtEpochMillis)
      ]);
    return query.watch();
  }

  Future<WorkoutFeedbackData?> feedbackForSessionNow(int sessionId) {
    final query = select(workoutFeedbacks)
      ..where((tbl) => tbl.sessionId.equals(sessionId))
      ..limit(1);
    return query.getSingleOrNull();
  }

  Future<void> upsert(WorkoutFeedbackData feedback) {
    return into(workoutFeedbacks).insertOnConflictUpdate(feedback);
  }
}
